colcon_bundle/verb/bundlefile.py

```diff
@@ -0,0 +1,107 @@
+        self.metadata = []
+        self.mode = mode
+
+    def add_metadata(self, path):
+        """
+        Add the json file at path to the metadata archive.
+
+        :param path: path to json file
+        """
+        self.metadata.append(path)
+
+    def add_overlay_archive(self, path):
+        """
+        Add the archive at path to the bundle as an overlay.
+
+        This does not write immediately, once the object is closed all
+        writes will occur.
+        """
+        self.overlays.append(path)
+
+    def close(self):
+        """Close the archive."""
+        max_data_size = 4 * 1024 * 1024
+        if 'w' in self.mode:
+            tempdir = tempfile.mkdtemp()
+            version_path = os.path.join(tempdir, 'version')
+
+            with open(version_path, 'w') as v:
+                v.write('2')
+            self.tarfile.add(version_path, arcname='version')
+
+            offset = max_data_size
+            overlay_metadata = []
+            for overlay in self.overlays:
+                name = os.path.basename(overlay)
+                checksum = filechecksum(overlay)
+                info = self.tarfile.gettarinfo(overlay, arcname=name)
+                header_size = len(info.tobuf())
+                file_size = os.stat(overlay).st_size
+                total = header_size + file_size
+                overlay_metadata.append({
+                    'name': name,
+                    'sha256': checksum,
+                    'offset': offset + header_size,
+                    'size': file_size
+                })
+                offset += total
+
+            metadata_path = os.path.join(tempdir, 'overlays.json')
+            with open(metadata_path, 'w') as md:
+                metadata = {
+                    'overlays': overlay_metadata
+                }
+                json.dump(metadata, md)
+
+            metadata_archive_path = os.path.join(tempdir, 'metadata.tar.gz')
+            with tarfile.open(metadata_archive_path, 'w:gz') as md:
+                md.add(metadata_path, arcname=os.path.basename(metadata_path))
+                for item in self.metadata:
+                    md.add(item, arcname=os.path.basename(item))
+            self.tarfile.add(metadata_archive_path,
+                             arcname=os.path.basename(metadata_archive_path))
+
+            if os.stat(metadata_archive_path).st_size > max_data_size:
+                raise RuntimeError('Metadata too large, must be less than 4MB')
+
+            tar_header_size = 512
+            pad_size = max_data_size - self.tarfile.offset - tar_header_size
+            pad_path = os.path.join(tempdir, 'pad')
+            with open(pad_path, 'wb') as f:
+                data = bytearray(pad_size)
+                f.write(data)
+            self.tarfile.add(pad_path, arcname='pad')
+
+            for overlay in self.overlays:
+                name = os.path.basename(overlay)
+                self.tarfile.add(overlay, arcname=name)
```

Comment:
Should we change the mode of the Bundle after this to closed? Should we close the tarfile here as well?

Reply:
We close() in the exit block. I'm not sure what the pattern is here, I'll check out the standard library.

Reply:
Yes, and yes.

Reply:
Used the same pattern that tar uses for checking closure and mode.